src/credentials/CHIPCert.cpp

```diff
@@ -708,20 +709,122 @@ CHIP_ERROR DetermineCertType(ChipCertificateData & cert)
 
 bool ChipDN::IsEqual(const ChipDN & other) const
 {
-    if (mAttrOID == kOID_Unknown || mAttrOID == kOID_NotSpecified || mAttrOID != other.mAttrOID)
-    {
+    // Same number of attributes?
+    if (mCount != other.mCount)
         return false;
+
+    // kOID_AttributeType_ChipNodeId has priority and is located at index 0, any other are at higher indices.
+    if (mAttrOID[0] == chip::ASN1::kOID_AttributeType_ChipNodeId)
+        return other.mAttrOID[0] == chip::ASN1::kOID_AttributeType_ChipNodeId &&
+            mAttrValue[0].mChipId == other.mAttrValue[0].mChipId;
+
+    for (int i = 1; i < mCount; ++i)
+    {
+        int j = 1;
+        for (; j < other.mCount; ++j)
+        {
+            if (mAttrOID[i] == other.mAttrOID[j])
+                break;
+        }
+
+	// No OID match?
+        if (j == other.mCount)
+            return false;
+
+        // OIDs matched, how about values?
+        if (IsChipX509Attr(mAttrOID[i]))
+        {
+            if (mAttrValue[i].mChipId != other.mAttrValue[j].mChipId)
+                return false;
+        }
+        else
+        {
+            if (mAttrValue[i].mString.mLen != other.mAttrValue[j].mString.mLen ||
+                memcmp(mAttrValue[i].mString.mValue, other.mAttrValue[j].mString.mValue, mAttrValue[i].mString.mLen) != 0)
+                return false;
+        }
     }
+    // All match
+    return true;
+}
 
-    if (IsChipIdX509Attr(mAttrOID))
+void ChipDN::Clear()
+{
+    for (int i = 0; i < RDN_NUM; ++i)
     {
-        return mAttrValue.mChipId == other.mAttrValue.mChipId;
+        mAttrOID[i] = chip::ASN1::kOID_NotSpecified;
     }
-    else
+    mCount = 0;
+}
+
+CHIP_ERROR ChipDN::Add(chip::ASN1::OID attrOID, uint64_t chipId)
+{
+    if (attrOID == chip::ASN1::kOID_NotSpecified || attrOID == chip::ASN1::kOID_Unknown)
+        return CHIP_ERROR_UNKNOWN_KEY_TYPE;
+
+    if (attrOID == chip::ASN1::kOID_AttributeType_ChipNodeId)
+    {
+        if (mAttrOID[0] == chip::ASN1::kOID_NotSpecified)
+        {
+            mAttrOID[0] = attrOID;
+            mCount++;
+        }
+        mAttrValue[0].mChipId = chipId;
+
+        return CHIP_NO_ERROR;
+    }
+
+    for (int i = 1; i < RDN_NUM; ++i)
+    {
+        if (mAttrOID[i] == chip::ASN1::kOID_NotSpecified)
+        {
+            mAttrOID[i]           = attrOID;
+            mAttrValue[i].mChipId = chipId;
+            mCount++;
```

Comment:
I don't think this behaves right, unless I am missing something.

Say I create a cert with a single RDN her, with a ChipCAId oid.  mCount will be 1.  The actual CA id will be stored in slot 1.  I create a second cert with a different ChipCAId value.  Its mCount will be 1 too.

If I now call IsEqual(), they will test equal, though I'd think they should not.

What am I missing?